lib/client.js

```diff
@@ -332,12 +331,42 @@ MatrixClient.prototype.uploadKeys = function(maxKeys, deferred) {
         }
 
         self._olmDevice.generateOneTimeKeys(numberToGenerate);
-        return _doKeyUpload(self);
+        return _uploadOneTimeKeys(self);
     });
 };
 
-// build the upload request, and return a promise which resolves to the response
-function _doKeyUpload(client) {
+// returns a promise which resolves to the response
+function _uploadDeviceKeys(client) {
+    if (!client._olmDevice) {
+        return q.reject(new Error("End-to-end encryption disabled"));
+    }
+
+    var userId = client.credentials.userId;
+    var deviceId = client.deviceId;
+
+    var deviceKeys = {
+        algorithms: client._cryptoAlgorithms,
+        device_id: deviceId,
+        keys: client._deviceKeys,
+        user_id: userId,
+        signatures: {},
```

Comment:
I don't think we should include the signature key here. 
https://matrix.org/docs/spec/server_server/unstable.html#signing-json